src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc

```diff
@@ -0,0 +1,190 @@
+    stream << "int32_t shift = " << shift << ";\n";
+    PrintIndent();
+    stream << "int32_t diff_min = " << diff_min << ";\n";
+    PrintIndent();
+    stream << cmsis_func_name << "(input,";
+    PrintIndent();
+    stream << " num_rows, row_size, mult, shift, diff_min, output);\n";
+    PrintIndent();
+    stream << "return;\n";
+  }
+};
+
+class CMSISNNModuleNode : public runtime::ModuleNode {
+ public:
+  CMSISNNModuleNode(const std::string& code, const std::string& fmt,
+                    const Array<String>& func_names)
+      : code_(code), fmt_(fmt), func_names_(func_names) {}
+
+  std::string GetSource(const std::string& format) final { return code_; }
+
+  const char* type_key() const { return "c"; }
+
+  PackedFunc GetFunction(const std::string& name, const ObjectPtr<Object>& sptr_to_self) final {
+    if (name == "get_symbol") {
+      return PackedFunc(
+          [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->func_names_[0]; });
+    } else if (name == "get_func_names") {
+      return PackedFunc(
+          [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->func_names_; });
+    } else {
+      return PackedFunc(nullptr);
+    }
+  }
+
+  void SaveToFile(const std::string& file_name, const std::string& format) final {
+    std::string fmt = GetFileFormat(file_name, format);
+    std::string meta_file = GetMetaFilePath(file_name);
+    if (fmt == "c" || fmt == "cu") {
+      ICHECK_NE(code_.length(), 0);
+      SaveBinaryToFile(file_name, code_);
+    } else {
+      ICHECK_EQ(fmt, fmt_) << "Can only save to format=" << fmt_;
+    }
+  }
+
+ protected:
+  std::string code_;
+  std::string fmt_;
+  Array<String> func_names_;
+};
+
+class CMSISNNModule : public Module {
+ public:
+  CMSISNNModule() {}
+  explicit CMSISNNModule(ObjectPtr<Object> n) : Module(n) {}
+  inline CMSISNNModuleNode* operator->();
+  inline const CMSISNNModuleNode* operator->() const;
+};
+
+inline CMSISNNModuleNode* CMSISNNModule::operator->() {
+  return static_cast<CMSISNNModuleNode*>(get_mutable());
+}
+
+static Module CMSISNNModuleNodeCreate(IRModule mod) {
+  bool output_ssa = false;
+  CodeGenCMSISNN cg;
+  Array<String> function_names;
+  cg.Init(output_ssa);
+  ICHECK(mod->functions.size() == 1) << "Supports modules with single PrimFunc.";
+  for (auto kv : mod->functions) {
+    ICHECK(kv.second->IsInstance<PrimFuncNode>()) << "CodegenCMSISNN: Can only take PrimFunc";
+    auto f = Downcast<PrimFunc>(kv.second);
+    auto global_symbol = f->GetAttr<String>(tvm::attr::kGlobalSymbol);
+    ICHECK(global_symbol.defined())
+        << "CodeGenCHost: Expect PrimFunc to have the global_symbol attribute";
+    function_names.push_back(global_symbol.value());
+    cg.AddFunction(f);
+  }
+  std::string code = cg.Finish();
+  auto n = make_object<CMSISNNModuleNode>(code, "c", function_names);
+  return Module(n);
+}
+
+TVM_REGISTER_GLOBAL("cmsisnn.create.runtime.module").set_body([](TVMArgs args, TVMRetValue* rv) {
```

Comment:
This namespace is bit confusing. Anyway, I dont think we would need this once the codegen from python disappears.

Reply:
I generally prefer to keep the namespace of c++.

In this case, it would be good : "runtime.CMSISNNModuleNodeCreate".